### PR TITLE
connect: deployments should wait for Connect sidecar checks

### PR DESCRIPTION
**Pull request by tgross**
When a Connect service is registered with Consul, Nomad includes the nested `Connect.SidecarService` field that includes health checks for the Envoy proxy. Because these are not part of the job spec, the alloc health tracker created by `health_hook` doesn't know to read the value of these checks.

In many circumstances this won't be noticed, but if the Envoy health check happens to take longer than the `update.min_healthy_time` (perhaps because it's been set low), it's possible for a deployment to progress too early such that there will briefly be no healthy instances of the service available in Consul.

Update the Consul service client to find the nested sidecar service in the service catalog and attach it to the results provided to the tracker. The tracker can then check the sidecar health checks.

Fixes: https://github.com/hashicorp/nomad/issues/19269

### CODE REVIEW

File: command/agent/consul/service_client.go [hunk 1]
```diff
@@ -1760,6 +1760,15 @@ func (c *ServiceClient) AllocRegistrations(allocID string) (*serviceregistration
 					sreg.Checks = append(sreg.Checks, check)
 				}
 			}
+
+			if sidecarService := services[serviceID+sidecarSuffix]; sidecarService != nil {
```
Comment (gulducat) [inline command/agent/consul/service_client.go:1764]: this feels like a bit of a mouthful, so I went to double-check it, and I found this thing:

https://github.com/hashicorp/nomad/blob/d9c9dc548a6037d9e71acdb8b717fd8784a1f581/command/agent/consul/service_client.go#L2123-L2130

nit-picky, but I think I prefer having the magic-string key encapsulated in that one spot:

```suggestion
			if sidecarService := getNomadSidecar(serviceID, services); sidecarService != nil {
```

**Review by gulducat — APPROVED** [review]
one lil suggest, but either way, LGTM!

File: command/agent/consul/service_client.go (hunk 1 above)
Comment (tgross) [reply command/agent/consul/service_client.go:1764]: Good catch :+1:

**Comment by tgross** [comment]
I'll add the backport labels on this once 1.7.0 GA goes out.

**Pull request merged**